testing/testing.go

```diff
@@ -0,0 +1,518 @@
+	Engine common.Engine
+	// DB is the database engine that the test case is running against.
+	// It provides access directly to Postgres, and has superuser access
+	// to the underlying database. If users want to execute ad-hoc queries,
+	// they should prefer to use the Engine, which parses Kwil's SQL standard,
+	// and guarantees determinism.
+	DB sql.DB
+	// Deployer is the public identifier of the user that deployed the schemas
+	// during test setup. It can be used to execute owner-only actions and procedures.
+	// To execute owner-only actions and procedures, set the Deployer to be the
+	// *common.ExecutionData.TransactionData.Signer field when executing the
+	// action/procedure.
+	Deployer []byte
+
+	// Logger is for logging information during execution of the test.
+	Logger Logger
+
+	// lastTxid is the last transaction ID that was used.
+	lastTxid []byte
+}
+
+// Txid returns a new, unused transaction ID.
+// It is deterministic, making tests repeatable.
+func (p *Platform) Txid() string {
+	if len(p.lastTxid) == 0 {
+		b := sha256.Sum256([]byte("first txid"))
+		p.lastTxid = b[:]
+		return hex.EncodeToString(b[:])
+	}
+
+	b := sha256.Sum256(p.lastTxid)
+	p.lastTxid = b[:]
+	return hex.EncodeToString(b[:])
+}
+
+// runWithPostgres runs the callback function with a postgres container.
+func runWithPostgres(ctx context.Context, opts *Options, fn func(context.Context, *pg.DB, Logger) error) (err error) {
+	if !opts.UseTestContainer {
+		db, err := pg.NewDB(ctx, &pg.DBConfig{
+			PoolConfig: pg.PoolConfig{
+				MaxConns: 11,
+				ConnConfig: pg.ConnConfig{
+					Host:   opts.Conn.Host,
+					Port:   opts.Conn.Port,
+					User:   opts.Conn.User,
+					Pass:   opts.Conn.Pass,
+					DBName: opts.Conn.DBName,
+				},
+			},
+		})
+		if err != nil {
+			return fmt.Errorf("error setting up database: %w", err)
+		}
+
+		defer db.Close()
+
+		return fn(ctx, db, opts.Logger)
+	}
+
+	// check if the user has docker
+	err = exec.Command(ctx, "docker")
+	if err != nil {
+		if strings.Contains(err.Error(), "not found") {
+			return fmt.Errorf("docker not found. Please ensure Docker is installed and running")
+		}
+		return fmt.Errorf("error checking for Docker installation: %w", err)
+	}
+
+	port := "52853"
+
+	// Run the container
+	bts, err := exec.CommandOutput(ctx, "docker", "run", "-d", "-p", fmt.Sprintf("%s:5432", port), "--name", "kwil-testing-postgres", "-e",
+		"POSTGRES_HOST_AUTH_METHOD=trust", "kwildb/postgres:latest")
+	if err != nil {
+		return fmt.Errorf("error running test container: %w", err)
+	}
+	defer func() {
+		err2 := exec.Command(ctx, "docker", "rm", "-f", "kwil-testing-postgres")
+		if err2 != nil {
+			if err == nil {
+				err = err2
+			} else {
+				err = errors.Join(err, err2)
+			}
+		}
+	}()
+
+	opts.Logger.Logf("running test container: %s", string(bts))
+
+	time.Sleep(1 * time.Second) // stupid hack needed for the container to be ready
```

Comment:
This is kind too hacky(not stable), maybe apply retry logic on  `pg.NewDB` if got connection error?